### DISCUSSION TITLE
Roadmap: Reusable React Component Library (Done. Only missing docs)

### DISCUSSION THREAD

**Discussion by denolfe** [payloadcms/payload#293] [Roadmap]
## Description

Export all Payload React components for re-use outside of Payload

## Implementation Detail

Use webpack to export all Payload React components, so a developer may use them elsewhere.

## Potential Breaking Changes

None

## Effort

High

**Comment by JarrodMFlesch** [comment]
We should attempt to convert SCSS vars to CSS vars where possible.

[Related discord thread](https://discord.com/channels/967097582721572934/1075154570147270787)

**Comment by jacobsfletch** [comment]
Related https://github.com/payloadcms/payload/discussions/2388

**Comment by miguelvictor** [comment]
Toast notifications please 💯 

[Related discord thread](https://discord.com/channels/967097582721572934/1089502682454167642)

**Comment by jacobsfletch** [comment]
Drawers 👇

https://github.com/payloadcms/payload/discussions/3042

**Comment by IAmNatch** [comment]
Form components would be great as well 🙏 
#3112

**Comment by wrobson-lllow** [comment]
Any updates to this? I would like to make a custom field built off the relationship where I can define how the options are sorted.

**Comment by jimzer** [comment]
I'm also working on plugins and it would be awesome to build the different UI components with Payload official components to match the look and feel 🔥

**Comment by jacobsfletch** [comment]
Storybook → https://github.com/payloadcms/payload/discussions/2165

**Comment by tyteen4a03** [comment]
Are there any movements to this?

**Reply by jacobsfletch** [reply]
@tyteen4a03 we're going to be starting work on this soon and will to have it finished by the end of Q4. We're going to be setting a new public roadmap in the coming days to bump this in priority.

**Comment by hatsumatsu** [comment]
This is great, can't wait to start building custom fields based on payload's core components!

One UI-related thing I noticed is the inconsistent style of icons. Maybe this is a something that could be fixed when working on the components? Using an existing set like https://remixicon.com could be a low-effort approach. I would love to see the great feature set of payload be reflected in the quality of UI 💅

<img width="373" alt="Screenshot 2023-12-22 at 10 44 45" src="https://github.com/payloadcms/payload/assets/3452400/dbe6d462-3162-4dec-b8f3-0219e404e807">
<img width="273" alt="Screenshot 2023-12-22 at 10 44 22" src="https://github.com/payloadcms/payload/assets/3452400/c6b51ef6-fa6e-4f83-acbe-b762219e5280">
<img width="327" alt="Screenshot 2023-12-22 at 10 46 39" src="https://github.com/payloadcms/payload/assets/3452400/d7ac6349-3cc6-45df-b9b2-f5bcd0c27e96">

**Reply by Jan-Eckerlein** [reply]
Wdym? Icons are the same. Hole components look different to reflect the different behaviour.
Only uploads are out of place. Don’t like em in general. And no link to the upload collection entry

**Reply by hatsumatsu** [reply]
@Jan-Eckerlein What I mean is that the different icons are inconsistent in style: compare the line width of the ❎ and 🔽 icon and the optical size of ❎ and ➕ . In the upload field the thin-lined "copy" icon sits next to an "edit" icon with a very different line width. It just doesn't look like these visual elements belong to the same design language. In the relation field this becomes even more obvious where the actual entry appears scaled down.

Using a pre-existing icon sets like https://remixicon.com that manage to deliver a consistent design language over a whole range of icons could be a low-hanging fruit in improving the UI of payload.

**Comment by jacobsfletch** [comment]
Hey everyone, quick update here. We're actively working on a new `@payloadcms/ui` package. This will go out together with [Payload 3.0](https://github.com/payloadcms/payload/discussions/4202) which is currently in alpha. In 3.0, the admin panel installs and consumes this package directly. Keep an eye on that thread as we work toward an official stable release. The `ui` packages is still in its early stages, but long term will include a full, standalone testing suite, Storybook, and more. I'll post more updates here as they roll in.

**Reply by wrobson-lllow** [reply]
Was this released with 3.0? is it compatible with 2.0?

**Reply by akhrarovsaid** [reply]
@wrobson-lllow Yep, you can install `@payloadcms/ui` into your v3 project right now. It's _not_ compatible with v2 unfortunately as it takes advantage of several new patterns and optimizations that are unique to v3.

**Comment by ahmadkakarr** [comment]
Great idea

**Comment by GermanJablo** [comment]
I'm going to mark this discussion as complete now that `@payloadcms/ui` is a thing. If there's anything missing, let me know!

**Reply by tyteen4a03** [reply]
I don't think this should be marked as done until the documentation are in place to support developers on *how* to use the components.

**Reply by GermanJablo** [reply]
You're right! I'm going to rename the discussion to reflect what's missing.

**Comment by RobinvdGriend** [comment]
Wondering what the status is on this! I've tried using components directly from `@payloadcms/ui`, but without documentation it has been very difficult. A lot of components do not have a very intuitive API and most props are undocumented. My experience is that it's not really possible to use the library without having a local clone of `@payloadcms/ui` to read implementation.

On a more abstract note, it's somewhat unclear right now on what is the intended way of customizing the admin or adding custom components. There's some documentation on using Tailwind, but using this to create components makes it hard to have a consistent styling and because the tailwind (base) classes are fighting with those defined by Payload the DX is not great. Using `@payloadcms/ui` would of course make sense, but is difficult right now because of the above reasons.

**Reply by jmcapra** [reply]
> There's some documentation on using Tailwind, but using this to create components makes it hard to have a consistent styling and because the tailwind (base) classes are fighting with those defined by Payload the DX is not great.

Spot on. This is my biggest gripe with Payload CMS right now, and probably one of the biggest efficiency drains for developing applications with Payload CMS. It seems the team has a total identity crisis regarding Tailwind vs SCSS. The current main options as a plugin or app developer include:

- Spend countless hours trying to reverse engineer the @payloadcms/ui components using type hints and the source code on github _(i.e, flatpack arrived with all the parts, but the instructions are in a foreign language)_
- Use tailwind without the base classes and just deal with that _(table and chairs have arrived, they're just missing the feet but it's okay - we can carve those out ourselves, technically)_
- Go full tailwind and spend a few minutes relishing in your beautiful new, meticulously-prompted and well-styled AI-generated UI component before you realise the sidebar and rest of the payload admin are totally wrecked and you need to start again _(the "throw batteries in the ocean" approach)_
- Roll your own component system from scratch, potentially go on a little [BEM sidequest](https://payloadcms.com/docs/admin/customizing-css) along the way but if you're anything like me, almost certainly waste a few days and do a worse job (less readable, more spaghetti/duplication, less performance) than if you had just used tailwind _(the "getting paid by the hour, not the job" approach)_

**Comment by dikaso** [comment]
The only thing we're missing is the storybook.

It would speed up development as well as contribute to further development of the components.
